core/nodefactory.lua

```diff
@@ -275,14 +287,27 @@ function nodefactory.discretionary:toText ()
 end
 
 function nodefactory.discretionary:outputYourself (typesetter, line)
-  -- TODO this is an indication of a deeper bug. If we were asked to output
-  -- discretionaries but the parent nnode is not hyphenated then we're
-  -- outputting things that were only used for measuring "what if" scenarios in
-  -- break point calculations. These nodes shouldn't exist at this point.
+  -- See typesetter:computeLineRatio() which implements the currently rather
+  -- messy hyphenated checks.
+  -- Example: consider the word "out-put-ter".
+  -- The node queue contains N(out)D(-)N(put)D(-)N(ter) all pointing to the same
+  -- parent N(output), and here we hit D(-)
+
+  -- Non-hyphenated parent: when N(out) was hit, we went for outputting
+  -- the whole parent, so all other elements must now be skipped.
   if self.parent and not self.parent.hyphenated then return end
+
+  -- It's possible not to have a parent (e.g. on a discretionary directly
+  -- added in the queue and not coming from the hyphenator logic).
```

Comment:
> a discretionary directly added...

Typically what shows up in test "bug-1528-discretionary" with using a `\discretionary` call. Only discretionary nodes from the Liang line breaking code have parents. (Again, if that matters for understanding?)